### ISSUE TITLE
Serialization format

### ISSUE THREAD

**Issue by Luthaf** [lab-cosmo/equistore-examples#2]
The serialization format defined in https://github.com/Luthaf/aml-storage-examples/blob/98175848e4c9c70353368621e8a6f029e3eedf1b/utils/io.py went through two iterations, both with advantages and drawbacks. I'm opening this to discuss which one we would like to use

- the first format was a collection of NetCDF-3 (classic format) files, grouped together in an uncompressed ZIP file. The files structure looked like this:
```
/  sparse.nc
   blocks-0.nc
   blocks-1.nc
     ...
   blocks-n.nc
```

Where each block contained the following NetCDF dimensions and variables:
```bash
# dimensions
samples
components
properties

sample_names
components_names
properties_names

name_length

gradient_<parameter>_sample
gradient_<parameter>_sample_names


# variables
values: double, (sample, components, properties)

samples: i32, (samples, sample_names)
sample_names: char, (sample_names, name_length)

components: i32, (components, components_names)
components_names: char, (components_names, name_length)

properties: i32, (properties, properties_names)
properties_names: char, (properties_names, name_length)

gradient_<parameter>: double (gradient_<parameter>_sample, components, properties)

gradient_<parameter>_samples: i32, (gradient_<parameter>_samples, gradient_<parameter>_sample_names)
gradient_<parameter>_sample_names: char, (gradient_<parameter>_sample_names, name_length)
```

---- 

The new and current format is based on HDF5, and the groups and datasets looks like this (using `@<XXX>` for datasets and `-> <XXX>` for attributes)

```
/   @sparse
        -> names
    blocks/
            0/
                @values
                @samples
                    -> names
                components/
                    @0
                        -> names
                    @1
                        -> names
                    ...
                    @n
                        -> names
                @properties
                    -> names
            
            gradients/
                <parameter>/
                    @data
                    @samples
                        -> names
                    components/
                        @0
                            -> names
                        @1
                            -> names
                        ...
                        @n
                            -> names
                    @properties
                        -> names

            1/
                ...
```

----

Overall, the HDF5-based format is much closer to the in-memory representation, and less hacky. The main issue is that the HDF5 library can be a pain to work with, and hard to install and use, for example if we use this format to store sparse/support/active points in GAP model that we want to load in LAMMPS.

The advantage of the ZIP + NetCDF-3 format is that both ZIP and NetCDF-3 are well defined formats with multiple implementations, meaning we can more easily integrate it directly within equistore without having to manage HDF-5 installations. The main issue is that it is not directly a standard format but a composition of two formats, meaning users have to write some more code to read/write it. It is also more hacky and further away from the in-memory representation.

Do you have any thoughts on this?

**Comment by ceriottm** [comment]
I think HDF5 is a pretty solid standard, isn't it? There are libraries installed everywhere, no?
I've seen this though https://towardsdatascience.com/the-best-format-to-save-pandas-data-414dca023e0d that seems to suggest HDF5 to be rather sluggish. 
Why do you think HDF5 will be harder to use with lammps than a custom mix of NetCDF and zip?

**Comment by Luthaf** [comment]
HDF5 is a pretty solid standard, and quite nice to use. But the format itself is quite hard to parse, and there is only one implementation, meaning we have to dynamically link to it with all related issues (library not found, library only present in one system, code compiled against one version but only an older one is available, etc.). This will increase the number of issues users will have to deal with (and bug reports we will have to deal with). It is also my understanding that HDF5 is not the most well-behaved library on this front.

NetCDF-3 isn't as nice to use from a developer point of view, but it has alternative implementations. The code previously used the implementation of the format in [scipy](https://docs.scipy.org/doc/scipy-1.7.1/reference/generated/scipy.io.netcdf_file.html), which is a pure Python implementation without any dependency on a hard to manage C library.

**Comment by PicoCentauri** [comment]
+1 for HDF5.

We could add a build option `build_own_hdf5` that downloads and installs exactly the version we need based on the major version of the storage library, feature generation library (rascaline) and our md interface library.

**Comment by Luthaf** [comment]
I had to battle against libhdf5 again today, so I'm more and more against using it: it will create a lot of installation issues for users/more work for us if we want to download & install a specific version of libhdf5.

Another format I though about that is (a) trivial to use from Python and (b) simple enough that custom parser exists in Rust/C/... is numpy's `.npz` files. 

These are ZIP archives containing `.npy` member files. We can use the same hierarchical idea as the HDF5 example, and the other piece of metadata we want to save (names of the Labels) can be stored as a custom dtype (the same dtype which is already used by the Python `Labels`) in the individual npy files.

I just pushed a prototype for this in `utils.io.read_npz/write_npz`: https://github.com/lab-cosmo/equistore-examples/blob/22784291892ea55f73ad5556fc5dce329d67203f/utils/io.py#L121-L198
